api/agent/protocol/json.go

```diff
@@ -35,54 +46,92 @@ func writeString(err error, dst io.Writer, str string) error {
 
 // TODO(xxx): headers, query parameters, body - what else should we add to func's payload?
 // TODO(xxx): get rid of request body buffering somehow
-func (h *JSONProtocol) DumpJSON(req *http.Request) error {
+func (h *JSONProtocol) DumpJSON(call *models.Call, req *http.Request) error {
 	stdin := json.NewEncoder(h.in)
 	bb := new(bytes.Buffer)
 	_, err := bb.ReadFrom(req.Body)
+	// todo: better/simpler err handling
 	if err != nil {
 		return err
 	}
+	// open
 	err = writeString(err, h.in, "{")
+	if err != nil {
+		return err
+	}
+
+	// call_id
+	err = writeString(err, h.in, `"call_id":`)
+	if err != nil {
+		return err
+	}
+	err = stdin.Encode(call.ID)
+	if err != nil {
+		return err
+	}
+
+	// body
+	err = writeString(err, h.in, ",")
 	err = writeString(err, h.in, `"body":`)
 	if err != nil {
 		return err
 	}
 	err = stdin.Encode(bb.String())
+	if err != nil {
+		return err
+	}
+
+	// request URL
 	err = writeString(err, h.in, ",")
-	err = writeString(err, h.in, `"headers":`)
+	err = writeString(err, h.in, `"request_url":`)
 	if err != nil {
 		return err
 	}
-	err = stdin.Encode(req.Header)
+	err = stdin.Encode(req.URL.String())
+	if err != nil {
+		return err
+	}
+
+	// headers
 	err = writeString(err, h.in, ",")
-	err = writeString(err, h.in, `"query_parameters":`)
+	err = writeString(err, h.in, `"headers":`)
 	if err != nil {
 		return err
 	}
-	err = stdin.Encode(req.URL.RawQuery)
-	err = writeString(err, h.in, "}")
+	err = stdin.Encode(req.Header)
+
+	// config
+	if call.EnvVars != nil && len(call.EnvVars) > 0 {
```

Comment:
route and app config available through env in full, why do we need them one more time here?

Reply:
Good point (same comment as @rdallman and I agree, should be in envs).